### PR TITLE
feat(source): add s3 source and csv parser

### PR DESCRIPTION
**Pull request by waruto210**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

- Add abstraction of file system data sources
- Add S3 source and csv parser

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Connector (sources & sinks)

### Release note

```SQL
create materialized source s(
    id int,
    name varchar,
    age int,
    primary key(id)
) with (
    connector = 's3',
    s3.region_name = 'ap-southeast-1',
    s3.bucket_name = 'mingchao-s3-source'
) row format csv [WITHOUT HEADER] delimited by ',';
```

## Refer to a related PR or issue link (optional)

#6511

### CODE REVIEW

File: src/connector/src/source/base.rs [hunk 1]
```diff
@@ -113,6 +117,32 @@ pub enum SplitReaderImpl {
...
+    pub fn into_stream(self) -> BoxFsSourceStream {
+        match self {
+            Self::S3(s3_reader) => s3_reader.into_stream(),
+        }
+    }
+
+    pub async fn create(
+        config: ConnectorProperties,
+        state: Vec<FsSplit>,
+        _columns: Option<Vec<Column>>,
```
[the remark this answers is not in the document]
Comment (waruto210) [reply src/connector/src/source/base.rs:134]: In fact, it seems that none of the existing split readers use this parameter.

[the remark this answers is not in the document]
Comment (tabVersion) [reply src/connector/src/source/base.rs:201]: > Why do we skip using columns?

only datagen connector uses this field

File: src/connector/src/source/base.rs [hunk 2]
```diff
@@ -169,6 +184,32 @@ pub enum SplitReaderImpl {
     GooglePubsub(Box<PubsubSplitReader>),
 }
 
+pub enum FsSplitReaderImpl {
```
Comment (yezizp2012) [inline src/connector/src/source/base.rs:187]: Why defining `FsSplitReaderImpl` in a separate enum? 🤔

Comment (waruto210) [reply src/connector/src/source/base.rs:187]: `FsSplitReaderImpl` impl a different reader trait.

**Review by tabVersion — APPROVED** [review]
generally LGTM

File: src/source/src/fs_connector_source.rs [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+#[derive(Clone, Debug)]
+pub struct FsConnectorSource {
+    pub config: ConnectorProperties,
+    pub columns: Vec<SourceColumnDesc>,
+    pub properties: HashMap<String, String>,
+    pub format: SourceFormat,
+    // for csv
+    pub csv_delimiter: u8,
+    pub csv_has_header: bool,
```
[the remark this answers is not in the document]
Comment (waruto210) [reply src/source/src/fs_connector_source.rs]: `ConnectorSource` and `FsConnectorSource` are responsible for building reader and parsing data, the `ConnectorSource` contains a parser directly, in a sense, it also contains parser config, the `FsConnectorSource` needs to create parser for each neww object/file, so I think it's reasonable to put the parser config here.

**Pull request merged**